main.nf

```diff
@@ -167,46 +486,708 @@ process fastqc {
+    input:
+    file(db) from file_kraken2_db
+
+    output:
+    set val("${db.baseName}"), file("${db.baseName}/*.k2d") into kraken2_database
+
+    script:
+    """
+    tar -xf "${db}"
+    """
+}
+
+trimmed_reads_kraken2
+    .combine(kraken2_database)
+    .set { kraken2_input }
+
+process kraken2 {
+    tag "${name}-${db_name}"
+    publishDir "${params.outdir}/Taxonomy/kraken2/${name}", mode: 'copy',
+            saveAs: {filename -> filename.indexOf(".krona") == -1 ? filename : null}
+
+    input:
+    set val(name), file(reads), val(db_name), file("database/*") from kraken2_input
+
+    output:
+    set val("kraken2"), val(name), file("results.krona") into kraken2_to_krona
+    file("kraken2_report.txt")
+
+    script:
+    def input = params.singleEnd ? "\"${reads}\"" :  "--paired \"${reads[0]}\" \"${reads[1]}\""
+    """
+    kraken2 \
+        --report-zero-counts \
+        --threads "${task.cpus}" \
+        --db database \
+        --fastq-input \
+        --report kraken2_report.txt \
+        $input \
+        > kraken2.kraken
+    cat kraken2.kraken | cut -f 2,3 > results.krona
+    """
+}
+
+process krona_db {
+    output:
+    file("taxonomy/taxonomy.tab") into file_krona_db
+
+    when:
+    ( params.centrifuge_db || params.kraken2_db ) && !params.skip_krona
+
+    script:
+    """
+    ktUpdateTaxonomy.sh taxonomy
+    """
+}
+
+centrifuge_to_krona
+    .mix(kraken2_to_krona)
+    .combine(file_krona_db)
+    .set { krona_input }
+
+process krona {
+    tag "${classifier}-${name}"
+    publishDir "${params.outdir}/Taxonomy/${classifier}/${name}", mode: 'copy'
+
+    input:
+    set val(classifier), val(name), file(report), file("taxonomy/taxonomy.tab") from krona_input
+
+    output:
+    file("*.html")
+
+    script:
+    """
+    ktImportTaxonomy "$report" -tax taxonomy
+    """
+}
+
+
+/*
+ * STEP 2 - Assembly
+ */
+process megahit {
+    tag "$name"
+    publishDir "${params.outdir}/", mode: 'copy',
+        saveAs: {filename -> filename.indexOf(".fastq.gz") == -1 ? "Assembly/$filename" : null}
+
+    input:
+    set val(name), file(reads) from trimmed_reads_megahit
+
+    output:
+    set val("MEGAHIT"), val("$name"), file("MEGAHIT/${name}.contigs.fa") into assembly_megahit_to_quast
+    set val("MEGAHIT"), val("$name"), file("MEGAHIT/${name}.contigs.fa") into assembly_megahit_to_metabat
+    file("MEGAHIT/*.log")
+
+    when:
+    !params.skip_megahit
+
+    script:
+    def input = params.singleEnd ? "-r \"${reads}\"" :  "-1 \"${reads[0]}\" -2 \"${reads[1]}\""
+    """
+    megahit -t "${task.cpus}" $input -o MEGAHIT --out-prefix "${name}"
+    """
+}
+
+
+/*
+ * metaSpades hybrid Assembly
+ */
+
+ files_lr_filtered
+    .combine(trimmed_sr_spadeshybrid, by: 0)
+    .set { files_pre_spadeshybrid }
+
+process spadeshybrid {
+    tag "$id"
+    publishDir "${params.outdir}/", mode: 'copy', pattern: "${id}*",
+        saveAs: {filename -> filename.indexOf(".fastq.gz") == -1 ? "Assembly/SPAdesHybrid/$filename" : null}
```

Comment:
I think that this should give the expected result - saving only the files you want to the results directory and giving them nice filenames. Without the need for a `cp` or `mv`.

```suggestion
        saveAs: {filename ->
            if (filename.indexOf(".fastq.gz") > 0) "$filename"
            else if (filename.indexOf("assembly_graph_with_scaffolds.gfa") > 0) "${id}_graph.gfa"
            else if (filename.indexOf("scaffolds.fasta") > 0) "${id}_scaffolds.fasta"
            else if (filename.indexOf("contigs.fasta") > 0) "${id}_contigs.fasta"
            else if (filename.indexOf("spades.log") > 0) "${id}_log.txt"
            else null
        }
```

Reply:
I had gotten your solution to work but then I have file name conflicts in the `busco_plot` process (cause I collect the assemblies there).

I'll go with `move` for now. I'm not saying make your own PR yet, but I'll delay solving this for some ulterior minor release 😄

Reply:
Hm, well I run into that problem before, maybe should have said this. I renamed this output files because of name conflicts downstream. Of course it would be possible to rename the input in the busco process but than the in the report the sample name would be messed up...

Reply:
Can you just rename the busco inputs using wildcards? See https://www.nextflow.io/docs/latest/process.html#multiple-input-files

Reply:
Oh sorry, you said @d4straub that you need the sample name for the MultiQC report.. Fair enough 🤷‍♂ 😄